### ISSUE TITLE
Using ZeroMQ instead or in parallel with bitcoind's blocknotify for DAO full nodes

### ISSUE THREAD

**Issue by KanoczTomas** [bisq-network/proposals#73]
> _This is a Bisq Network proposal. Please familiarize yourself with the [submission and review process](https://docs.bisq.network/proposals.html)._
# Using ZeroMQ instead or in parallel with bitcoind's blocknotify for DAO full nodes

## Current state
Currently the bisq `DAO` full node requires a blocknotify script that sends the hash of the newest block to a port defined in a script.
Current bitcoin.conf entry needed for blocknotify:
```
blocknotify=./blocknotify %s
```
Example blocknotify script content (Linux):
```
#!/bin/bash
echo $1 | nc -w 1 127.0.0.1 5120
```
A script needs to exist, have proper execution rights for bitcoind to successfully run it. Besides user needs to have netcat installed which might be cumbersome on other platforms than Linux.

## Why is it a good idea to change current state ?
As the blocknotify script uses netcat (nc) in the current state it is essentially using a tcp connection to localhost. While the script used with blocknotify can be altered to run the bisq node and bitcoind deamon on a separate physical (or virtual) server a better approach using ZeroMQ is available.

## What is ZeroMQ?
`ZeroMQ` is a subscription service that requires very little load on the server and client side. The server connects to a tcp socket where it emits events. The client can subscribe to those events when connected to the tcp socket of the server. The server writes data only once, yet all clients get it.

## bitcoind ZeroMQ support
Bitcoind supports `ZeroMQ` for a long time. There are 4 events (topics) a subscriber can subscribe to:
  1. **hashtx** - emits the hash of the newest transaction
  2. **hashblock** - emits the hash of the newest block
  3. **rawtx** - emits the full newest raw transaction 
  4. **rawblock** - emits the full newest raw block

`ZeroMQ` is considered a notification service by bitcoind and `RPC` has to be further used to retrieve data and have 100% guarantee the data is correct. It means it is not a replacement for `RPC`, but another notification option.
`ZeroMQ` has to be activated in the configuration of `bitcoind`, an example where all events are emmited to the same port (the clients filter based on topic to which they subscribed):
```
zmqpubhashtx=tcp://127.0.0.1:28332
zmqpubhashblock=tcp://127.0.0.1:28332
zmqpubrawblock=tcp://127.0.0.1:28332
zmqpubrawtx=tcp://127.0.0.1:28332
```
## Example usage in other project
An example of how the `rawtx` is used to retrieve fees and show `mempool transactions` in realtime (**warning, JavaScript code, do not look at it if you hate JS**):
https://github.com/KanoczTomas/01Cnode/blob/master/server.js#L84

## What needs to change in bisq to use ZeroMQ?
A `java` library can be used to connect to `bitcoind`'s `ZeroMQ` service and listen for **hashblock** events. The code handling the blockhash does not need to change, as the `ZeroMQ` message payload contains the same data as netcat is sending in the current implementation (namely a string = hash of the newest block). 
Perhaps this `java` lib could be used: https://github.com/zeromq/jeromq
Equivalent `ZeroMQ` configuration in `bitcoind` using port 5120:
```
zmqpubhashblock=tcp://127.0.0.1:5120
```
## What bisq gains by using ZeroMQ?
Besides dropping the need to have netcat installed and a script distributed (per OS of course) to handle notifications I can imagine the bisq full node to create a `ZeroMQ` server as well and expose `ZeroMQ` events as `bitcoind` has for the `DAO`. That way a lot of possibilities for monitoring tools, block explorers towards bisq could be opened using very little code and effort. 
Suggested **event**(topic) names for the bisq `DAO`:
  1. **bsqhashtx** - emits the hash of the newest BSQ transaction
  2. **bsqhashblock** - emits the hash of the newest block containing BSQ transactions
  3. **bsqrawtx** - emits the full raw BSQ transaction 
  4. **bswrawblock** - emits the full raw block containing BSQ transactions

**Comment by sqrrm** [comment]
I like the idea of using zeromq as I have good experiences using it previously. The benefit of replacing the blocknotify script is not huge, but still seems better and perhaps we could even listen to rawblock and parse that directly if there is a library for it. I don't see this as anything urgent but nice to have.

The benefits to a block explorer and monitoring tools seem rather interesting. Currently the entire daostate has to be reparsed for the explorer as there is a plain dump of the bsq state to disk, that could obviously be improved on. For monitoring it would probably be really good, enabling a lot more fine grained data.

As manfred mentioned during the call though, I agree this is not urgent, but nice to have.

**Comment by ManfredKarrer** [comment]
The option for using it as API for other services (monitor, explorer) sounds interesting. For the DAO use case I don't really see a strong benefit which would justify the effort to change it and I somehow like the simplicity of the netcat approach which seems to carry little risk due little complexity. But I am not against it if any developer would like to work on that, I just don't see a priority in it.

But as said for other use cases like mentioned above it sounds quite interesting. Maybe @freimair can have a look if it could be useful for monitoring?

**Comment by cbeams** [comment]
Just having a look at this proposal now at @KanoczTomas's request. The idea sounds good in general. I'm familiar with ZeroMQ and Bitcoin Core's use of it, and it makes sense to me that we'd have similar support, especially for the explorer. I'd say for this work to get priority we should have someone owning it from the explorer side, and ideally someone that can own it end-to-end by implementing the ZeroMQ service on the Java side as well (@sqrrm seems the obvious candidate here, but I'll let him speak for himself). Per @KanoczTomas's question in Keybase chat, I don't see a _runtime_ dependency on the in-development gRPC API work in the sense that the ZeroMQ service would call into our gRPC services, but I do see a more general dependency, in that it would be good to ship the at least the initial cut of the gRPC work first, along with the new Bisq daemon, because it will be easier and more straightforward how to implement the ZeroMQ service once that is done. Plus I have a number of changes that have not been merged yet that any ZeroMQ work in parallel would likely cause conflicts with.

**Comment by mpolavieja** [comment]
I am closing this proposal as stalled.  If anyone thinks it should be reopened, please comment.
